pkg/controller/container-runtime-config/container_runtime_config_controller.go

```diff
@@ -710,6 +706,63 @@ func (ctrl *Controller) syncImageConfig(key string) error {
 	return nil
 }
 
+func registriesConfigIgnition(templateDir string, controllerConfig *mcfgv1.ControllerConfig, role string,
+	insecureRegs, blockedRegs []string, icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) (*igntypes.Config, error) {
+	var registriesTOML []byte
+	if insecureRegs != nil || blockedRegs != nil || len(icspRules) != 0 {
+		// Generate the original registries config
+		_, _, originalRegistriesIgn, err := generateOriginalContainerRuntimeConfigs(templateDir, controllerConfig, role)
+		if err != nil {
+			return nil, fmt.Errorf("could not generate origin ContainerRuntime Configs: %v", err)
+		}
+		dataURL, err := dataurl.DecodeString(originalRegistriesIgn.Contents.Source)
+		if err != nil {
+			return nil, fmt.Errorf("could not decode original registries config: %v", err)
+		}
+		registriesTOML, err = updateRegistriesConfig(dataURL.Data, insecureRegs, blockedRegs, icspRules)
+		if err != nil {
+			return nil, fmt.Errorf("could not update registries config with new changes: %v", err)
+		}
+	}
+	registriesIgn := createNewRegistriesConfigIgnition(registriesTOML)
+	return &registriesIgn, nil
+}
+
+// RunImageBootstrap generates MachineConfig objects for mcpPools that would have been generated by syncImageConfig,
+// except that mcfgv1.Image is not available.
+func RunImageBootstrap(templateDir string, controllerConfig *v1.ControllerConfig, mcpPools []*v1.MachineConfigPool, icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) ([]*v1.MachineConfig, error) {
+	// ImageConfig is not available.
+	insecureRegs := []string(nil)
+	blockedRegs := []string(nil)
+
+	var res []*v1.MachineConfig
+	for _, pool := range mcpPools {
+		role := pool.Name
+		managedKey := getManagedKeyReg(pool)
+
+		registriesIgn, err := registriesConfigIgnition(templateDir, controllerConfig, role,
+			insecureRegs, blockedRegs, icspRules)
+		if err != nil {
+			return nil, err
+		}
+
+		tempIgnCfg := ctrlcommon.NewIgnConfig()
+		mc := mtmpl.MachineConfigFromIgnConfig(role, managedKey, &tempIgnCfg)
+		mc.Spec.Config = *registriesIgn
+		// Explicitly do NOT set GeneratedByControllerVersionAnnotationKey so that the first run of the non-bootstrap controller
+		// always rebuilds registries.conf (with the insecureRegs/blockedRegs values actually available).
+		mc.ObjectMeta.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion: apicfgv1.SchemeGroupVersion.String(),
+				Kind:       "Image",
+				// Name and UID is not set, the first run of syncImageConfig will overwrite these values.
```

Comment:
I have no idea whether this is safe.

Reply:
what I'm worried about here is that we'll end up with a rendered machineconfig which is different than the one produced by the in-cluster sync-loop. This only means 1 thing, this is not going to work cause there will be a drift if we can't generate exactly the same thing we're generating at bootstrap

Reply:
we may want the ImageConfig to be available at bootstrap as well so we can make sure this is going to generate the same stuff - will think more about this.

Reply:
This code intentionally does not set `GeneratedByControllerVersionAnnotationKey` to anything. That should guarantee that when the MCC starts, `syncImageConfig` will _always_ update the object on first start.

Or are you saying that such an update would break things?

(I’m also assuming that future upgrades of the cluster don’t use the bootstrap process, so the bootstrap code does not need to worry about any pre-existing `getManagedKeyReg(pool)`-named objects.)

Reply:
> Or are you saying that such an update would break things?

the state today is:

- the bootstrap process generates a given `rendered-<pool>-<hash>` MC
- the sync loop generates a given `rendered-<pool>-<hash>` MC
- both have to match or it's not going to work

I'm not fully sure the above is the case (too tired now lol). But I suspect that if this code doesn't produce the same MachineConfig (cause it's missing ImageConfig) than what syncImageConfig would produce, then we'll have a drift and it's not gonna work. Will check again better later today.

Reply:
WRT making `ImageConfig` available, I agree that it would be _clean_ to generate exactly the same configuration on the bootstrap path, but it’s not just `ImageConfig`: `getValidRegistries` requires `ClusterVersion`.**`Status.Desired.Image`**, and looking at
https://github.com/openshift/installer/blob/580b58fa7ef8c8904f65392830aaf18ede2b3ddd/data/data/manifests/bootkube/cvo-overrides.yaml.template and https://github.com/openshift/cluster-version-operator/blob/398afc117e3d418664572cf2b9554ce9fcd2bcc8/bootstrap/bootstrap-pod.yaml , this is not set by the installer, but only by a bootstrap pod — that runs at about the time of the MCC bootstrap, but it’s not available at the MCO bootstrap time, and presumably (?) the MCC bootstrap can’t use the Kubernetes API to read (and block on) the `ClusterVersion.Status` to be available.

Reply:
> * the bootstrap process generates a given `rendered-<pool>-<hash>` MC
> * the sync loop generates a given `rendered-<pool>-<hash>` MC
> * both have to match or it's not going to work

That happens in the render controller, and AFAICS that should not be affected; this only generates one more MC for the render controller to merge.

Yes, it could be possible that the `registries.conf` input to the render controller is the same during bootstrap and after bootstrap — in that case `syncImageConfig` would create an updated MC, but that would not trigger the render controller to create a new `rendered-*` MC. That seems correct to me in principle.

(Currently, the first post-bootstrap `syncImageConfig` always triggers the render controller to re-render because `syncImageConfig` creates a `registries.conf` MC, even if that MC has no `Spec.Config.Storage.Files` (the `registriesTOMLConfig == nil` case); and because `MergeMachineConfigs` does not exclude no-files empty configs, this causes the hash to change. So, if anything in the render code differed between the bootstrap path and the controller, and silently expected the first post-bootstrap run to trigger a re-render, yes, that could be a problem. Comparing `syncGeneratedMachineConfig` and `RunBootstrap`, it does seem that the bootstrap code does not set `pool.Spec.Configuration.Source`, and does set `pool.Status.Configuration.Name`, so there might be that kind of a difference. That could be problematic for `operator.isMachineConfigPoolConfigurationValid` and `syncRequiredMachineConfigPools`, but that’s too deep a rabbit hole for me right now.)

Reply:
… I mean, if you are worried only about the hash, we can always add a `# Dummy comment to ensure the registries.conf is different on bootstrap` line to the bootstrap one. But that is a badly targeted hack.

Another way to argue this is that the render controller exists longer than `syncImageConfig`, so _if_ the render controller worked correctly before `syncImageConfig` was added, then the extra `registries.conf` MC that currently triggers a re-render was never required.  OTOH the render controller is not _that_ much older than `syncImageConfig`, so I can’t quickly tell for sure.

Reply:
> This code intentionally does not set `GeneratedByControllerVersionAnnotationKey` to anything. That should guarantee that when the MCC starts, `syncImageConfig` will _always_ update the object on first start.

Oops, I didn’t notice that the MCC bootstrap only saves the results of rendering, not the intermediate MCs; those exist only in memory. Well, one more reason for `syncImageConfig` to always be used to produce the copy of record.

(This does not affect the discussion about the rendered MC being correct/consistent AFAICS.)

Reply:
has this been tested somehow on a cluster? I want to avoid merging something which has not been manually tested at least if there's a way to test this out

Reply:
@runcom i'm going to try to test it this afternoon if i can figure out how to build the images, get them into a release image, and mirror it